app/controllers/time_slot_controller.rb

```diff
@@ -52,40 +51,22 @@ def cancel
   end
 
   def index
-    @appointment = current_patient.current_appointment
-    @ubs = @appointment.try(:ubs)
-
-    @patient_can_schedule = current_patient.can_schedule?
-
     return render_vaccinated if current_patient.vaccinated?
 
-    @gap_in_days = slot_params[:gap_in_days].to_i || 0
-    @current_day = Time.zone.now + @gap_in_days.days
-
-    @time_slots = {}
-
     last_appointment = current_patient.last_appointment
     return if last_appointment&.second_dose? && @current_day.to_date < last_appointment.start.to_date
 
-    Ubs.where(active: true).each do |ubs|
```

Comment:
Instead of iterating through Ubs's and then through appointments, we can do it all with a single SQL query.

Reply:
Agreed. We could add such queries to the model and test them, leaving the controller with very few logic.